File: Next Smallest Palindrome.java
class Solution {
    static int[] nextPalindrome(int[] num) {
        // code here
        int n = num.length;
        boolean all9 = true;
        for (int d : num){
            if (d!=9){
                all9 = false;
                break;
            }
        }
        if (all9){
            int[] res = new int[n+1];
            res[0] = 1;
            res[n] = 1;
            return res;
        }
        int mid = (n + 1) / 2;
        int[] left = new int[mid];
        for (int i = 0; i < mid; i++){
            left[i] = num[i];
        }
        int[] pal = makePalindrome(left, n);
        if (compare(pal, num) > 0){
            return pal;
        }
        for (int i = mid - 1; i >= 0; i--){
            if (left[i] < 9){
                left[i]++;
                break;
            } else{
                left[i] = 0;
            }
        }
        return makePalindrome(left, n);
    }
    static int[] makePalindrome(int[] left, int n){
        int[] res = new int[n];
        int mid = (n + 1) / 2;
        for (int i = 0; i < mid; i++){
            res[i] = left[i];
        }
        int i = mid - 1;
        int j = n % 2 == 0? mid : mid - 1;
        while (i >= 0 && j < n){
            res[j] = left[i];
            i--;
            j++;
        }
        return res;
    }
    static int compare(int[] a, int[] b){
        for (int i = 0; i < a.length; i++){
            if (a[i] > b[i]) return 1;
            if (a[i] < b[i]) return -1;
        }
        return 0;
    }
    
}

File: Next Smallest Palindrome.py
class Solution:
    def nextPalindrome(self, num):
        # code here
        n = len(num)
        if num == [9] * n:
            return [1] + [0] * (n - 1) + [1]
        arr = num[:]
        mid = n // 2
        i = mid - 1
        j = mid + 1 if n % 2 else mid
        while i >= 0:
            arr[j] = arr[i]
            i -= 1
            j += 1
        if arr <= num:
            i = mid-1
            j = mid + 1 if n % 2 else mid
            if n % 2 == 1:
                if arr[mid] < 9:
                    arr[mid] += 1
                else:
                    arr[mid] = 0
                    carry = 1
                    while i >= 0:
                        if arr[i] < 9:
                            arr[i] += 1
                            carry = 0
                            break
                        else:
                            arr[i] = 0
                        i -= 1
                    i = mid - 1
                    j = mid + 1
                    while i >= 0:
                        arr[j] = arr[i]
                        i -= 1
                        j += 1
            else:
                carry = 1
                while i >= 0 and carry:
                    if arr[i] < 9:
                        arr[i] += 1
                        carry = 0
                    else:
                        arr[i] = 0
                    i -= 1
                i = mid - 1
                j = mid 
                while i >= 0:
                    arr[j] = arr[i]
                    i -= 1
                    j += 1
        return arr
